theano/tensor/nnet/ctc_wrapper.c

```diff
@@ -0,0 +1,264 @@
+    create_flat_labels( in_labels, &(context->flat_labels), &(context->label_lengths) );
+
+    if ( ( NULL == context->label_lengths ) || ( NULL == context->flat_labels ) )
+    {
+        // Destroy previous CTC context before returning exception
+        ctc_context_destroy( context );
+
+        PyErr_Format( PyExc_MemoryError,
+            "Could not allocate storage for labels and their lengths" );
+        return 1;
+    }
+
+    npy_int minibatch_size = PyArray_DIMS( in_activations )[1];
+    npy_int alphabet_size = PyArray_DIMS( in_activations )[2];
+
+    npy_float32 * costs = NULL;
+    npy_intp cost_size = minibatch_size;
+
+    if ( (*out_costs) == NULL ||                       // Symbolic variable has no memory backing
+         PyArray_NDIM( *out_costs ) != 1 ||            // or, matrix has the wrong size
+         PyArray_DIMS( *out_costs )[0] != cost_size )
+    {
+        Py_XDECREF( *out_costs );
+        // Allocate new matrix
+        *out_costs = (PyArrayObject *) PyArray_ZEROS( 1, &cost_size, NPY_FLOAT32, 0 );
+
+        if ( NULL == (*out_costs) )
+        {
+            // Destroy previous CTC context before returning exception
+            ctc_context_destroy( context );
+
+            PyErr_Format( PyExc_MemoryError,
+                "Could not allocate storage for CTC costs" );
+            return 1;
+        }
+    }
+
+    costs = (npy_float32 *) PyArray_DATA( *out_costs );
+
+    npy_float32 * gradients = NULL;
+
+    if ( NULL != out_gradients )  // If gradient computation is not disabled
+    {
+        if ( NULL == (*out_gradients) ||  // Symbolic variable has no real backing
+            PyArray_NDIM( *out_gradients ) != 3 ||
+            PyArray_DIMS( *out_gradients )[0] != PyArray_DIMS( in_activations )[0] ||
+            PyArray_DIMS( *out_gradients )[1] != PyArray_DIMS( in_activations )[1] ||
+            PyArray_DIMS( *out_gradients )[2] != PyArray_DIMS( in_activations )[2] )
+        {
+            // Existing matrix is the wrong size. Make a new one.
+            // Decrement ref counter to existing array
+            Py_XDECREF( *out_gradients );
+            // Allocate new array
+            *out_gradients = (PyArrayObject *) PyArray_ZEROS(3, PyArray_DIMS( in_activations ),
+                NPY_FLOAT32, 0);
+
+            if ( NULL == (*out_gradients) )
+            {
+                // Destroy previous CTC context before returning exception
+                ctc_context_destroy( context );
+
+                PyErr_Format( PyExc_MemoryError,
+                    "Could not allocate storage for CTC gradients!" );
+                return 1;
+            }
+        }
+        gradients = (npy_float32 *) PyArray_DATA( *out_gradients );
+    }
+
+    size_t cpu_workspace_size;
+    int ctc_error;
+
+    ctc_error = ctc_check_result( get_workspace_size( context->label_lengths,
+        context->input_lengths, alphabet_size, minibatch_size, context->options,
```

Comment:
Well, now `context->label_lengths` and `context->input_lengths` are `npy_int` arrays, ie. `npy_int32`(4 bytes). and `get_workspace_size()` expects them to be `int` arrays ( https://github.com/baidu-research/warp-ctc/blob/master/include/ctc.h#L123 ), but an `int` can be 2, 4 or even 8 bytes depending on hardware, operating system and compiler.

To avoid any future problems, I suggest than `context->label_lengths`, `context->input_lengths` and `context->flat_labels` be all `int`, so that they will be always passed as array of current `int` to CTC functions:

```C
typedef struct ctc_context {
    struct ctcOptions options;
    void * workspace;
    int * input_lengths; // int
    int * flat_labels; // int
    int * label_lengths; // int
} ctc_context_t;
```

Then, in functions `create_contiguous_input_lengths()` and `create_flat_labels()`, when you extract values from PyArrayObject, you cast PyArrayObject data to its real pointer and then you convert the value you got to an int.

```C
void create_contiguous_input_lengths( PyArrayObject * input_lengths_arr, /* int */ int ** input_lengths )
{
    npy_int num_elements = PyArray_DIMS( input_lengths_arr )[0];

    // Allocate an array of int, not npy_int
    *input_lengths = (int *) malloc( num_elements * sizeof(npy_int) );

    if ( NULL == (*input_lengths) )
        return;

    for( npy_int elem_idx = 0; elem_idx < num_elements; ++elem_idx )
    {
        // This could stay as-is, as value will be implicitely casted to int.
        (*input_lengths)[elem_idx] = *( (npy_int *) PyArray_GETPTR1( input_lengths_arr, elem_idx ) );
    }
}

void create_flat_labels( PyArrayObject * label_matrix, /*int*/ int ** flat_labels,
    /*int*/ int ** label_lengths )
{
    
    // ...
    
    // Allocate an array of int, not npy_int
    *flat_labels = (int *) malloc( rows * cols * sizeof(npy_int) );
    //...

    // Allocate an array of int, not npy_int
    *label_lengths = (int *) malloc( rows * sizeof(npy_int) );

    // ...
            
            // This could stay as-is, as value will be implicitely casted to int.
            npy_int label = *( (npy_int *) PyArray_GETPTR2( label_matrix, row_idx, col_idx ) );
            
            // ...

}
```

So, we keep the types expected by CTC API, but we do conversions when we extract values from PyArrayObjects.

The last thing to do is to make sure that we know the types of PyArrayObjects (see my remark in `make_node()`).